### PR TITLE
[WIP] Added upload task to class TransferManager.

### PR DESCRIPTION
**Pull request by qdbgw3ihnjw**
Added member function addUploadTask(..), getUploadTask(..), cancelUpload(..), startUploadTask(..)  in class TransferManager.

Added class TransferProgress for processing transfer rate.

### CODE REVIEW

File: src/filebrowser/transfer-mgr.cpp [hunk 1]
```diff
@@ -123,19 +379,831 @@ void TransferManager::cancelAllDownloadTasks()
...
+             zql = sqlite3_mprintf(
+                "REPLACE INTO FileTasks (account_signature, "
+                "type, repo_id, path, local_path, status, parent_task ) "
+                "VALUES (%Q, %Q, %Q, %Q, %Q, 'pending', %Q) ",
+                toCStr(account_signature_),
+                toCStr(task_type),
+                toCStr(repo_id),
+                toCStr(repo_full_path),
+                toCStr(local_path),
+                toCStr(QString::number(folder_task_id)));
```
Comment (lins05) [inline src/filebrowser/transfer-mgr.cpp]: Change `%Q` to `%d`, then you don't need `toCStr(QString::number(folder_task_id)))`

File: src/filebrowser/file-browser-dialog.cpp [hunk 2]
```diff
@@ -794,70 +833,87 @@ void FileBrowserDialog::onDownloadFinished(bool success)
...
+    const QFileInfo file = QFileInfo(local_path);
+    const QString file_name = file.fileName();
+    const SeafDirent dirent = SeafDirent::file(file_name, static_cast<quint64>(file.size()));
+    table_model_->replaceItem(file_name, dirent);
+}
+
+void FileBrowserDialog::onUploadFailed(const FileTaskRecord* task)
+{
+    QString msg = tr("Failed to upload file: %1").arg(task->fail_reason);
+    seafApplet->warningBox(msg, this);
```
Comment (lins05) [inline src/filebrowser/file-browser-dialog.cpp]: If the upload is canceled by the user, then we should not show a "task cancelled" error dialog.

File: src/filebrowser/transfer-mgr.cpp [hunk 3]
```diff
@@ -123,19 +379,831 @@ void TransferManager::cancelAllDownloadTasks()
...
+    } else {
+        return false;
+    }
+}
+
+void TransferManager::cancelTransfer(const QString& repo_id,
+                                     const QString& path)
+{
+    cancelDownload(repo_id, path);
+    cancelUpload(repo_id, path);
```
Comment (lins05) [inline src/filebrowser/transfer-mgr.cpp]: This is not good code style. Be clear about the logic: either cancel a download, or cancel a upload.

File: src/filebrowser/file-table.cpp [hunk 4]
```diff
@@ -745,16 +746,18 @@ void FileTableView::onMove()
...
 }
 
-void FileTableView::onCancelDownload()
+void FileTableView::onCancelTransfer()
 {
     if (item_ == NULL) {
         const QList<const SeafDirent*> dirents = getSelectedItemsFromSource();
         for (int i = 0; i < dirents.size(); i++) {
-            emit cancelDownload(*dirents[i]);
+            source_model_->removeItemNamed(dirents[i]->name);
```
Comment (lins05) [inline src/filebrowser/file-table.cpp]: Only remove the item if a upload is cancelled (not download).

File: src/filebrowser/file-table.cpp [hunk 5]
```diff
@@ -745,16 +746,18 @@ void FileTableView::onMove()
...
+            emit cancelTransfer(*dirents[i]);
         }
-        return;
+    } else {
+        source_model_->removeItemNamed(item_->name);
```
Comment (lins05) [inline src/filebrowser/file-table.cpp]: Same here.

**Comment by sunyuzhao** [comment]
1, If the uploaded file is renamed with an existing file, Existing files will also  have a progress bar display, and upload failed.
2, including the task of uploading the folder will fail.

**Pull request closed without merging**